core/src/main/scala/iliad/gl/renderGraph.scala

```diff
@@ -0,0 +1,600 @@
+
+  def validate(g: Graph.Constructed): ValidatedNel[String, Unit] =
+    (nodesUnique *> linksUnique *> nodesConnected *> endNodesOnScreen).apply(g)
+}
+
+object GraphInstantiation {
+  import RenderModel._
+
+  type RValidated[A] = ReaderT[ValidatedNel[String, ?], Graph.Instance, A]
+
+  private def textures(n: Draw.Instance): ValidatedNel[String, Unit] =
+    n.piped.uniforms.traverseUnit {
+      case (name, cons) =>
+        n.uniforms.get(name) match {
+          case Some(u: Texture.Instance) =>
+            if (u.constructor == cons) ().valid
+            else
+              s"Uniform texture $name has inconsistent texture instance $u for node $n".invalidNel
+          case Some(u: Texture.Image) =>
+            s"Uniform texture $name is image $u instead of texture instance $cons for $n".invalidNel
+          case None =>
+            s"No texture provided for uniform $name of node $n".invalidNel
+        }
+    }
+
+  private def images(n: Draw.Instance): ValidatedNel[String, Unit] =
+    n.imageNames.traverseUnit { name =>
+      n.uniforms.get(name) match {
+        case Some(u: Texture.Image) => ().valid
+        case Some(u: Texture.Instance) =>
+          s"Uniform image $name is texture instead of image for node $n".invalidNel
+        case None =>
+          s"No image provided for uniform $name of node $n".invalidNel
+      }
+    }
+
+  private def attributes(n: Draw.Instance): ValidatedNel[String, Unit] =
+    n.constructor.program.vertex.attributes.traverseUnit { a =>
+      if (n.model.model.vertex.ref.buffer.attributes.contains(a)) ().valid
+      else s"Attribute $a is not present for node $n".invalidNel
+    }
+
+  private def filledNodes(ns: List[Node.Instance]): RValidated[Unit] =
+    ReaderT { g =>
+      val unfilled = g.constructed.links.flatMap { l =>
+        if (ns.exists(_.constructor == l.start) && !ns.exists(
+                _.constructor == l.end)) {
+          Some(l.end)
+        } else if (ns.exists(_.constructor == l.end) && !ns.exists(
+                       _.constructor == l.start)) {
+          Some(l.start)
+        } else None
+      }
+      if (unfilled.nonEmpty)
+        s"instance has unfilled nodes ${unfilled}".invalidNel
+      else ().valid
+    }
+
+  private def lift(v: ValidatedNel[String, Unit]): RValidated[Unit] =
+    KleisliExtra.lift(v)
+
+  private def checks(ns: List[Node.Instance])
+    : ReaderT[Xor[NonEmptyList[String], ?], Graph.Instance, Unit] = {
+    val vs = lift(ns.filterClass[Draw.Instance].traverseUnit(attributes)) *>
+        lift(ns.filterClass[Draw.Instance].traverseUnit(images)) *>
+        lift(ns.filterClass[Draw.Instance].traverseUnit(textures)) *>
+        filledNodes(ns)
+
+    vs.transform(
+        new (ValidatedNel[String, ?] ~> Xor[NonEmptyList[String], ?]) {
+      def apply[A](v: ValidatedNel[String, A]): NonEmptyList[String] Xor A =
+        v.toXor
+    })
+  }
+
+  def put(ns: List[Node.Instance])
```

Comment:
We'll need data verification elsewhere, to ensure that textures and models are loaded